### PR TITLE
 Added option to read warning attributes from config, file or url 

### PR DESCRIPTION
**Pull request by BaranekD**

### CODE REVIEW

File: lib/WarningConfigurationConfig.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    public function getSourceOfWarningAttributes()
+    {
+        $config = Configuration::getConfig(self::CONFIG_FILE_NAME);
+        return $config;
+    }
+
+    public function getWarningAttributes()
+    {
+        $warningIsOn = false;
```
Comment (pajavyskocil) [inline lib/WarningConfigurationConfig.php]: Please move attributes initialization(line 30-33) into abstract class WarningConfiguration

File: lib/WarningConfigurationFile.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+/**
+ * Implementation of WarningConfiguration using json file as the source of warning attributes
+ * @package SimpleSAML\Module\perun
+ * @author Dominik Baránek <0Baranek.dominik0@gmail.com>
+ */
+class WarningConfigurationFile
+{
+    const CONFIG_FILE_NAME = 'module_perun.php';
+    const WARNING_FILE = 'disco.warning.file';
+    const WARNING_IS_ON = 'disco.warning.isOn';
```
Comment (pajavyskocil) [inline lib/WarningConfigurationFile.php]: Constants on lines 18-21 not used

File: lib/WarningConfigurationFile.php [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    {
+        $warningIsOn = false;
+        $warningUserCanContinue = true;
+        $warningTitle = '';
+        $warningText = '';
+
+        $data = self::getSourceOfWarningAttributes();
+
+        if (!is_null($data)) {
+            if (isset($data['warningIsOn'])) {
```
Comment (pajavyskocil) [inline lib/WarningConfigurationFile.php]: Please use constant for 'warningIsOn', 'warningUserCanContinue', 'warningTitle', 'warningText'

File: lib/WarningConfigurationUrl.php [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+/**
+ * Implementation of WarningConfiguration using json url as the source of warning attributes
+ * @package SimpleSAML\Module\perun
+ * @author Dominik Baránek <0Baranek.dominik0@gmail.com>
+ */
+class WarningConfigurationUrl
+{
+    const CONFIG_FILE_NAME = 'module_perun.php';
+    const WARNING_URL = 'disco.warning.url';
+    const WARNING_IS_ON = 'disco.warning.isOn';
```
Comment (pajavyskocil) [inline lib/WarningConfigurationUrl.php]: The same as for file WarningConfigurationFile

File: lib/WarningConfigurationUrl.php [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                "perun:WarningConfigurationUrl: missing or invalid disco.warning.url parameter in module_perun.php"
+            );
+        }
+
+        return $attributes;
+    }
+
+    public function getWarningAttributes()
+    {
+        $warningIsOn = false;
```
Comment (pajavyskocil) [inline lib/WarningConfigurationUrl.php]: The same as for file WarningConfigurationFile

File: lib/WarningConfigurationConfig.php [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        return $config;
+    }
+
+    public function getWarningAttributes()
+    {
+        $data = self::getSourceOfWarningAttributes();
+
+        if (!is_null($data)) {
+            try {
+                $this->warningIsOn = $data->getBoolean(self::WARNING_IS_ON);
```
Comment (dBucik) [inline lib/WarningConfigurationConfig.php]: Use getBoolean(self:WARNING_IS_ON, false) instead. With this approach, you can get rid of the catch block, as the line will not throw any exception and the default value will be set (as on line 40)

File: lib/WarningConfigurationConfig.php [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+                    "perun:WarningConfigurationConfig: " .
+                    "missing or invalid disco.warning.isOn parameter in module_perun.php"
+                );
+                $this->warningIsOn = false;
+            }
+        }
+
+        if ($this->warningIsOn) {
+            try {
+                $this->warningUserCanContinue = $data->getBoolean(self::WARNING_USER_CAN_CONTINUE);
```
Comment (dBucik) [inline lib/WarningConfigurationConfig.php]: The same situation as with getting "WARNING_IS_ON", use method with two parameters instead and get rid of the catch block.

File: lib/WarningConfigurationFile.php [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        }
+
+        return $data;
+    }
+
+    public function getWarningAttributes()
+    {
+        $data = self::getSourceOfWarningAttributes();
+
+        if (!is_null($data)) {
```
Comment (dBucik) [inline lib/WarningConfigurationFile.php]: replace _!is_null($data)_ with _null !== $data_

File: lib/WarningConfigurationUrl.php [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        }
+
+        return $attributes;
+    }
+
+    public function getWarningAttributes()
+    {
+        $data = self::getSourceOfWarningAttributes();
+
+        if (!is_null($data)) {
```
Comment (dBucik) [inline lib/WarningConfigurationUrl.php]: replace _!is_null($data)_ with _null !== $data_

File: themes/perun/perun/disco-tpl.php [hunk 10]
```diff
@@ -30,89 +30,48 @@
...
-try {
-    $configPerun = Configuration::getConfig(PERUN_CONFIG_FILE_NAME);
+    $config = Configuration::getConfig(CONFIG_FILE_NAME);
 } catch (\Exception $ex) {
-    Logger::warning("perun:disco-tpl: invalid module_perun.php file");
+    Logger::warning("perun:disco-tpl: missing or invalid module_perun.php config file");
 }
 
-if (!is_null($configPerun)) {
+if (!is_null($config)) {
```
Comment (dBucik) [inline themes/perun/perun/disco-tpl.php]: replace _!is_null($config)_ with _null !== $config_

File: themes/perun/perun/disco-tpl.php [hunk 11]
```diff
@@ -30,89 +30,48 @@
...
     try {
-        $addInstitutionUrl = $configPerun->getString(ADD_INSTITUTION_URL);
+        $addInstitutionUrl = $config->getString(ADD_INSTITUTION_URL);
     } catch (\Exception $ex) {
         Logger::warning("perun:disco-tpl: missing or invalid addInstitution.URL parameter in module_perun.php file");
     }
 }
 
-if (!is_null($configPerun)) {
+if (!is_null($config)) {
```
Comment (dBucik) [inline themes/perun/perun/disco-tpl.php]: replace _!is_null($config)_ with _null !== $config_

File: lib/WarningConfigurationFile.php (hunk 2 above)
Comment (BaranekD) [reply lib/WarningConfigurationFile.php]: Done

File: lib/WarningConfigurationUrl.php (hunk 4 above)
Comment (BaranekD) [reply lib/WarningConfigurationUrl.php]: Done

File: lib/WarningConfigurationConfig.php [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        $config = Configuration::getConfig(self::CONFIG_FILE_NAME);
+        return $config;
+    }
+
+    public function getWarningAttributes()
+    {
+        $data = self::getSourceOfWarningAttributes();
+
+        if (!is_null($data)) {
```
[the remark this answers is not in the document]
Comment (BaranekD) [reply lib/WarningConfigurationConfig.php]: Done

File: lib/WarningConfigurationConfig.php (hunk 6 above)
Comment (BaranekD) [reply lib/WarningConfigurationConfig.php]: Done

File: lib/WarningConfigurationConfig.php (hunk 7 above)
Comment (BaranekD) [reply lib/WarningConfigurationConfig.php]: Done

File: lib/WarningConfigurationFile.php (hunk 8 above)
Comment (BaranekD) [reply lib/WarningConfigurationFile.php]: Done

File: lib/WarningConfigurationUrl.php (hunk 9 above)
Comment (BaranekD) [reply lib/WarningConfigurationUrl.php]: Done

File: themes/perun/perun/disco-tpl.php (hunk 10 above)
Comment (BaranekD) [reply themes/perun/perun/disco-tpl.php]: Done

File: themes/perun/perun/disco-tpl.php (hunk 11 above)
Comment (BaranekD) [reply themes/perun/perun/disco-tpl.php]: Done

File: lib/Disco.php [hunk 13]
```diff
@@ -107,6 +107,18 @@ public function handleRequest()
             HTTP::redirectTrustedURL($url);
         }
 
+        try {
+            $warningInstance = WarningConfiguration::getInstance();
+            $warningAttributes = $warningInstance->getWarningAttributes();
+        } catch (Exception $ex) {
+            $warningAttributes = array(
```
Comment (pajavyskocil) [inline lib/Disco.php]: Please use [] instead of array()

File: lib/WarningConfiguration.php [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    /**
+     * @return string data with warning attributes
+     */
+    abstract public function getSourceOfWarningAttributes();
+
+    /**
+     * @return array with warning attributes
+     */
+    abstract public function getWarningAttributes();
+}
```
Comment (pajavyskocil) [inline lib/WarningConfiguration.php]: Please add new function to check if warningAttributes are correct.

**Pull request merged**